part_1/solution.py

```diff
@@ -0,0 +1,1327 @@
+    """
+    Logs a batch of images on jupyter using ipywidget.
+
+    Args:
+        batch (dict): A dictionary containing the batch of images to be logged.
+
+    Returns:
+        None
+    """
+    batch_phase = batch["source"][:, :, 0, :, :]  # batch_size x z_size x Y x X tensor.
+    batch_size = batch_phase.shape[0]
+    batch_membrane = batch["target"][:, 1, 0, :, :].unsqueeze(
+        1
+    )  # batch_size x 1 x Y x X tensor.
+    batch_nuclei = batch["target"][:, 0, 0, :, :].unsqueeze(
+        1
+    )  # batch_size x 1 x Y x X tensor.
+
+    p1, p99 = np.percentile(batch_membrane, (0.1, 99.9))
+    batch_membrane = np.clip((batch_membrane - p1) / (p99 - p1), 0, 1)
+
+    p1, p99 = np.percentile(batch_nuclei, (0.1, 99.9))
+    batch_nuclei = np.clip((batch_nuclei - p1) / (p99 - p1), 0, 1)
+
+    p1, p99 = np.percentile(batch_phase, (0.1, 99.9))
+    batch_phase = np.clip((batch_phase - p1) / (p99 - p1), 0, 1)
+
+    plt.figure()
+    fig, axes = plt.subplots(
+        batch_size, n_channels, figsize=(n_channels * 2, batch_size * 2)
+    )
+    [N, C, H, W] = batch_phase.shape
+    for sample_id in range(batch_size):
+        axes[sample_id, 0].imshow(batch_phase[sample_id, 0])
+        axes[sample_id, 1].imshow(batch_nuclei[sample_id, 0])
+        axes[sample_id, 2].imshow(batch_membrane[sample_id, 0])
+
+        for i in range(n_channels):
+            axes[sample_id, i].axis("off")
+            axes[sample_id, i].set_title(dataset.channel_names[i])
+    plt.tight_layout()
+    plt.show()
+
+
+# %%
+# Initialize the data module.
+
+BATCH_SIZE = 4
+
+# 5 is a perfectly reasonable batch size
+# (batch size does not have to be a power of 2)
+# See: https://sebastianraschka.com/blog/2022/batch-size-2.html
+
+data_module = HCSDataModule(
+    data_path,
+    z_window_size=1,
+    architecture="UNeXt2_2D",
+    source_channel=["Phase3D"],
+    target_channel=["Nucl", "Mem"],
+    split_ratio=0.8,
+    batch_size=BATCH_SIZE,
+    num_workers=8,
+    yx_patch_size=(256, 256),  # larger patch size makes it easy to see augmentations.
+    augmentations=[],  # Turn off augmentation for now.
+    normalizations=[],  # Turn off normalization for now.
+)
+data_module.setup("fit")
+
+print(
+    f"Samples in training set: {len(data_module.train_dataset)}, "
+    f"samples in validation set:{len(data_module.val_dataset)}"
+)
+train_dataloader = data_module.train_dataloader()
+
+# Instantiate the tensorboard SummaryWriter, logs the first batch and then iterates through all the batches and logs them to tensorboard.
+
+writer = SummaryWriter(log_dir=f"{log_dir}/view_batch")
+# Draw a batch and write to tensorboard.
+batch = next(iter(train_dataloader))
+log_batch_tensorboard(batch, 0, writer, "augmentation/none")
+writer.close()
+
+
+# %% [markdown]
+# If your tensorboard is causing issues, you can visualize directly on Jupyter /VSCode
+
+# %%
+log_batch_jupyter(batch)
+
+# %% [markdown] tags=[]
+# <div class="alert alert-info">
+#
+# ### Task 1.3
```

Comment:
split the questions from the actual tasks, and use lists for the tasks so they know they’ve done everything